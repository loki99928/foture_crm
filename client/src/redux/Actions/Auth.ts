import {InferActionTypes} from "../store";
import {initialStateAuth} from "../Reducer/Auth";

export const actions = {
    setUserData: (userId: string | undefined, email: string | undefined) => ({
        type: 'CRM/AUTH/SET_USER_DATA',
        data: {userId, email}
    } as const),
    getUserToken: (accessToken: string | undefined, refreshToken: string | undefined) => ({
        type: 'CRM/AUTH/GET_USER_TOKEN',
        data: {accessToken, refreshToken}
    } as const),
    toggleIsAuth: (isAuth: boolean) => ({type: 'CRM/AUTH/IS_AUTH', isAuth} as const),
    toggleIsLoad: () => ({type: 'CRM/AUTH/IS_LOAD'} as const),
    statusUserTokenNewPassword: (token: string) => ({
        type: 'CRM/AUTH/CHANGE_USER_TOKEN_NEW_PASSWORD',
        data: {token}
    } as const)
}

export type InitialStateType = typeof initialStateAuth
export type ActionTypeAuth = InferActionTypes<typeof actions>
