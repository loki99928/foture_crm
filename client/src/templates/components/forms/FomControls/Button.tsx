import React from "react";
import cn from "classnames";
import s from "./Form.module.scss";

export const Button: React.FC<any> = props => {
    const {disabled, ...rest} = props
  return (
      <div>
          <button className={cn(s.formBtn, s.formBtn_hover, s.formBtn_active, s.formBtn_focus)}
                  disabled={disabled}
                  type="submit"
                  {...rest}
          >
              submit
          </button>
      </div>
  )
}