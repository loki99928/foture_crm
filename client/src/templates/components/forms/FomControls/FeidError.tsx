import cn from "classnames";
import React from "react";
import s from "./Form.module.scss";

export const FieldError: React.FC<any> = (props: any) => {

    const {message, ...rest} = props

    return (
        <div className={cn(s.formTextError, s.formServer__error)} data-testid="formTextError">
            {message}
        </div>
    )
};