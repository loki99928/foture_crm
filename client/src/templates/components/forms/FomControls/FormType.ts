export const RULE = {
    EMAIL_MAX_LENGTH: 50,
    EMAIL_SPECIAL_CASE: /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/,
    PASSWORD_MIN_LENGTH: 5,
    PASSWORD_MAX_LENGTH: 50,
    PASSWORD_LOWERCASE: /^(?=.*[a-z])/,
    PASSWORD_UPPERCASE: /^(?=.*[A-Z])/,
    PASSWORD_NUMBER: /^(?=.*[0-9])/,
    PASSWORD_SPECIAL_CASE: /^(?=.*[!@#$%^&*])/
}
export const MESSAGE = {
    EMAIL_REQUIRED: 'Required',
    EMAIL_TO_SHORT: 'Too Short. Min length 5 symbol',
    EMAIL_TO_LONG: 'Too Long. Max length 50 symbol',
    EMAIL_IS_NOT_VALID: 'Email is not valid',
    PASSWORD_REQUIRED: 'Required',
    PASSWORD_MIN_LENGTH: 'Too Short. Min length 5 symbol',
    PASSWORD_MAX_LENGTH: 'Too Long. Max length 50 symbol',
    PASSWORD_SPECIAL_CASE: 'Must One Special Case Character(@, #, $, %, ^ ,&, *)',
    PASSWORD_MISMATCH: 'Password mismatch',
    PASSWORD_LOWERCASE: 'Must one Lowercase',
    PASSWORD_UPPERCASE: 'Must one Uppercase',
    PASSWORD_NUMBER: 'Must One Number',
}

type Inputs = {
    example: string,
    exampleRequired: string,
};

export interface fieldsForm {
    email: string
    password: string
}

export type FormikType = {
    setFieldError: (field: string, errorMsg: string) => void
    setSubmitting: (isSubmitting: boolean) => void
}