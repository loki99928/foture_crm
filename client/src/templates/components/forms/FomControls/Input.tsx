import React from "react";
import cn from "classnames"

import s from "./Form.module.scss"

const Input: React.FC<any> = props => {
    const {handlers, errors, field, label, ...rest} = props

    const error = errors[props['field']]

    const value = handlers.watch(field)

    return (
        <div
            className={cn(s.formBlockField, s.formBlock__field, {[s.formFieldError]: error}, {[s.formFieldApproved]: !error && value?.length > 0})}>
            <input {...handlers.register(field)}
                   className={cn(s.formField)}
                   data-testid={`input_${field}`}
                   autoComplete="on"
                   {...rest}
            />
            {error && (
                <div className={cn(s.formTextError, s.form__error)} data-testid="formTextError">
                    <div className={cn(s.formTextError, s.formServer__error)}>
                        {error.message}
                    </div>
                </div>
            )}
            <div className={cn(s.formLabel, {[s.formLabelStatic]: value?.length > 0})}>
                <label htmlFor="email">{label}</label>
            </div>
        </div>
    )
}

export default Input