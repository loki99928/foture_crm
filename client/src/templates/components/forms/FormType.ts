export type FormikType = {
    setFieldError: (field: string, errorMsg: string) => void
    setSubmitting: (isSubmitting: boolean) => void
}