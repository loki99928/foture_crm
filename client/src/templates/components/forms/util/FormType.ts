export const RULE = {
    EMAIL_MAX_LENGTH: 50,
    EMAIL_SPECIAL_CASE: /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/,
    PASSWORD_MIN_LENGTH: 5,
    PASSWORD_MAX_LENGTH: 50,
    PASSWORD_LOWERCASE: /^(?=.*[a-z])/,
    PASSWORD_UPPERCASE: /^(?=.*[A-Z])/,
    PASSWORD_NUMBER: /^(?=.*[0-9])/,
    PASSWORD_SPECIAL_CASE: /^(?=.*[!@#$%^&*])/,
    NAME_MIN_LENGTH: 5,
    NAME_MAX_LENGTH: 50,
}

export const MESSAGE = {
    EMAIL_REQUIRED: 'Required',
    EMAIL_TO_SHORT: 'Too Short. Min length 5 symbol',
    EMAIL_TO_LONG: 'Too Long. Max length 50 symbol',
    EMAIL_IS_NOT_VALID: 'Email is not valid',
    PASSWORD_REQUIRED: 'Required',
    PASSWORD_MIN_LENGTH: 'Too Short. Min length 5 symbol',
    PASSWORD_MAX_LENGTH: 'Too Long. Max length 50 symbol',
    PASSWORD_SPECIAL_CASE: 'Must One Special Case Character(@, #, $, %, ^ ,&, *)',
    PASSWORD_MISMATCH: 'Password mismatch',
    PASSWORD_LOWERCASE: 'Must one Lowercase',
    PASSWORD_UPPERCASE: 'Must one Uppercase',
    PASSWORD_NUMBER: 'Must One Number',
    NAME_MIN_LENGTH: 'Too Short. Min length 5 symbol',
    NAME_MAX_LENGTH: 'Too Long. Max length 50 symbol',
}

export interface fieldsFormAuth {
    email: string
    password: string
}

export interface fieldsFormRegister {
    email: string
    password: string
}

export interface fieldsFormForget {
    email: string
}

export interface fieldsFormNewPassword {
    password: string
    double_password: string
}

export interface fieldsSearchForm {
    search: string
}

export interface fieldsUserSettingsForm {
    email: string
    password: string
    double_password: string
    login: string
}
export type FormikType = {
    setFieldError: (field: string, errorMsg: string) => void
    setSubmitting: (isSubmitting: boolean) => void
}