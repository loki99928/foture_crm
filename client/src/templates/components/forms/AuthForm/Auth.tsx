import React from "react";
import {useDispatch, useSelector} from "react-redux";
import {NavLink} from "react-router-dom";
import cn from "classnames"
import {SubmitHandler, useForm} from "react-hook-form";
import * as Yup from "yup";
import {yupResolver} from "@hookform/resolvers/yup";

import {TUser} from "../../../../redux/reducer/auth/types";
import {getIsLoad, getMessage} from "../../../../redux/reducer/auth/selectors";
import {actionsAuth} from "../../../../redux/reducer/auth/actions";
import {useClearForm} from "../../../../hooks/useClearForm";
import s from "../FomControls/Form.module.scss"
import bannerForm from "../../../assets/images/bg-head-form.jpg"
import {fieldsForm} from "../FomControls/FormType";
import Input from "../FomControls/Input";
import {Button} from "../FomControls/Button";
import {validate} from "../FomControls/Validate";
import {FieldError} from "../FomControls/FeidError";

const SignupSchema = Yup.object().shape({
    email: validate.email,
    password: validate.password
});

const Auth: React.FC = () => {

    useClearForm()

    const {handleSubmit, formState: {errors, isValid}, ...handlers} = useForm<fieldsForm>({
        // defaultValues: {
        //     email: 'loki99928@yandex.ru',
        //     password: '123Qw@',
        // },
        mode: "all",
        resolver: yupResolver(SignupSchema)
    });

    const dispatch = useDispatch()
    const message = useSelector(getMessage)
    const isLoad = useSelector(getIsLoad)

    const onSubmit: SubmitHandler<fieldsForm> = (data: TUser) => {
        dispatch(actionsAuth.authUserRequest(data))
    };

    return (
        <div className={s.blockForm}>
            <div className={s.formBanner}>
                <img src={bannerForm} alt=""/>
            </div>
            <div className={s.containerFields}>
                <h2 className={s.formTitle}>Authorize Info</h2>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <Input errors={errors}
                           field='email'
                           handlers={handlers}
                           type='text'
                           label='Your email'/>
                    <Input errors={errors}
                           field='password'
                           handlers={handlers}
                           type='password'
                           label='Your password'/>
                    {message && <FieldError message={message}/>}
                    <Button disabled={isLoad || !isValid}/>
                </form>
                <div className={cn(s.formFooter, s.footer__form)}>
                    <NavLink to="/registration/">registration</NavLink>
                    <NavLink to="/forget/">forget password</NavLink>
                </div>
            </div>
        </div>
    );
}
export default Auth