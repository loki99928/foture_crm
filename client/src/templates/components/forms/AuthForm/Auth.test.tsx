import React from "react";
import {act, fireEvent, screen} from "@testing-library/react";

import Auth from "./";
import {renderWithRouter} from "../../../../helpers/test/renderWithRouter";
import {MESSAGE} from "../FomControls/FormType";

describe('AuthForm', () => {

    describe('input email', () => {

        it('have', function () {
            renderWithRouter(<Auth/>)
            expect(screen.queryByTestId('input_email')).toBeInTheDocument()
        })

        it('required', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            await act(async () => {
                // userEvent.
                fireEvent.blur(email)
            });
            expect(screen.getByTestId("test_email")).toBeInTheDocument();
            expect(screen.getByTestId("test_email")).toHaveTextContent(MESSAGE.EMAIL_REQUIRED);
        })

        it('validate', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            await act(async () => {
                fireEvent.change(email, {
                    target: {value: 'loki99928@yandex.ru'}
                })
            });
            expect(screen.queryByTestId("test_email")).toBeNull();
        })

        it('not validate', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            await act(async () => {
                fireEvent.change(email, {
                    target: {value: 'loki99928yandex'}
                })
            });
            expect(screen.queryByTestId("test_email")).toBeInTheDocument();
            expect(screen.queryByTestId("test_email")).toHaveTextContent(MESSAGE.EMAIL_IS_NOT_VALID);
        })

        it('too long', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            await act(async () => {
                fireEvent.change(email, {
                    target: {value: 'loki999282342234234234346ty546547456456546/jkdxjlmnkfvbcvb@yandex.ru'}
                })
            });
            expect(screen.queryByTestId("test_email")).toBeInTheDocument();
            expect(screen.queryByTestId("test_email")).toHaveTextContent(MESSAGE.EMAIL_TO_LONG);
        })
    })

    describe('input password', () => {

        it('have', () => {
            renderWithRouter(<Auth/>)
            expect(screen.getByTestId('input_password')).toBeInTheDocument()
        })

        it('validate', async () => {
            renderWithRouter(<Auth/>)
            const password = screen.getByTestId('input_password')
            await act(async () => {
                fireEvent.change(password, {
                    target: {value: '123123aA@'}
                })
            });
            expect(screen.queryByTestId("test_password")).toBeNull();
        })

        it('too short', async () => {
            renderWithRouter(<Auth/>)
            const password = screen.getByTestId('input_password')
            await act(async () => {
                fireEvent.change(password, {
                    target: {value: 'lk'}
                })
            });
            expect(screen.queryByTestId("test_password")).toBeInTheDocument();
            expect(screen.queryByTestId("test_password")).toHaveTextContent(MESSAGE.PASSWORD_MIN_LENGTH);
        })

        it('too long', async () => {
            renderWithRouter(<Auth/>)
            const password = screen.getByTestId('input_password')
            await act(async () => {
                fireEvent.change(password, {
                    target: {value: '1232133554756786788566345345345345456756878989008907890678'}
                })
            });
            expect(screen.queryByTestId("test_password")).toBeInTheDocument();
            expect(screen.queryByTestId("test_password")).toHaveTextContent(MESSAGE.PASSWORD_MAX_LENGTH);
        })
    })

    describe('btn Send', () => {
        it('have btn Send', () => {
            renderWithRouter(<Auth/>)
            expect(screen.getByRole('button', {name: /Send/i})).toBeInTheDocument()
        })
        it('btn disabled after loaded page', async () => {
            renderWithRouter(<Auth/>)
            const btn = screen.getByRole('button', {name: /Send/i})
            expect(btn).toBeDisabled()
        })
        it('disabled after error field of email', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            const btn = screen.getByRole('button', {name: /Send/i})
            await act(async () => {
                fireEvent.blur(email)
            });
            expect(btn).toBeDisabled()
        })
        it('not disabled after the error disappears', async () => {
            renderWithRouter(<Auth/>)
            const email = screen.getByTestId('input_email')
            const password = screen.getByTestId('input_password')
            const btn = screen.getByRole('button', {name: /Send/i})
            await act(async () => {
                fireEvent.blur(email)
                fireEvent.change(email, {
                    target: {value: 'loki99928@yandex.ru'}
                })
                fireEvent.change(password, {
                    target: {value: '123123aA@'}
                })
            });
            expect(btn).not.toBeDisabled()
        })
        it('disabled after error field of password', async () => {
            renderWithRouter(<Auth/>)
            const password = screen.getByTestId('input_password')
            const btn = screen.getByRole('button', {name: /Send/i})
            await act(async () => {
                fireEvent.blur(password)
            });
            expect(btn).toBeDisabled()
        })
    })

    // todo-dv нужно разобраться как тестировать переход по ссылкам
    describe('link', () => {
        describe('registration', () => {
            it('to have', () => {
                renderWithRouter(<Auth/>)
                expect(screen.getByRole('link', {name: /registration/i}).closest('a')).toHaveAttribute('href', '/registration/')
            })
        })

        describe('forget', () => {
            it('to have', () => {
                renderWithRouter(<Auth/>)
                expect(screen.getByRole('link', {name: /forget password/i}).closest('a')).toHaveAttribute('href', '/forget/')
            })
        })
    })
})