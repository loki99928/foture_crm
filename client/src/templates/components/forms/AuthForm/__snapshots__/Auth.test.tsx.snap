// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm Snapshot 1`] = `
<DocumentFragment>
  <div
    class="blockForm"
  >
    <form
      action="#"
    >
      <div
        class="formBanner"
      >
        <img
          alt=""
          src="bg-head-form.jpg"
        />
      </div>
      <div
        class="containerFields"
      >
        <h2
          class="formTitle"
        >
          Authorize Info
        </h2>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_email"
            for="email"
            id="email"
            name="email"
            type="email"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="email"
            >
              Your email
            </label>
          </div>
        </div>
        <div
          class="formBlockField formBlock__field"
        >
          <input
            autocomplete="on"
            class="formField"
            data-testid="input_password"
            for="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="formLabel"
          >
            <label
              for="password"
            >
              Your password
            </label>
          </div>
        </div>
        <div
          class="formBlockCheck formBlock__check"
        >
          <div>
            <label
              class="formContainerCheck"
              for="remember"
            >
              <input
                class="formFieldCheck"
                id="remember"
                name="remember"
                type="checkbox"
                value="false"
              />
              <span
                class="formCheckStyle form__check"
              />
              remember me
            </label>
          </div>
        </div>
        <div>
          <button
            class="formBtn formBtn_hover formBtn_active"
            type="submit"
          >
            Send
          </button>
        </div>
        <div
          class="formFooter footer__form"
        >
          <a
            class=""
            href="/registration/"
          >
            registration
          </a>
          <a
            class=""
            href="/forget/"
          >
            forget password
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
